nameserver/server_test.go

```diff
@@ -112,6 +116,76 @@ func TestDNSServer(t *testing.T) {
 	// needs two servers, each listening on its own address
 }
 
+func TestTCPDNSServer(t *testing.T) {
+	const (
+		port          = 17625
+		localTestName = "test1.weave.local."
+		testAddr1     = "10.0.2.1"
+	)
+	dnsAddr := fmt.Sprintf("localhost:%d", port)
+
+	InitDefaultLogging(true)
+	var zone = new(ZoneDb)
+
+	// Insert a bunch of IPs for the same name
+	ansCount := 0
+	for i := 0; i < 255; i++ {
+		for j := 0; j < 10; j++ {
+			ip := net.ParseIP(fmt.Sprintf("10.0.%d.%d", i, j))
+			ansCount += 1
+			err := zone.AddRecord(containerID, localTestName, ip)
+			wt.AssertNoErr(t, err)
+		}
+	}
+
+	// Run another DNS server for fallback
+	s, fallbackAddr, err := RunLocalUDPServer("127.0.0.1:0")
+	wt.AssertNoErr(t, err)
+	defer s.Shutdown()
+
+	_, fallbackPort, err := net.SplitHostPort(fallbackAddr)
+	wt.AssertNoErr(t, err)
+
+	config := &dns.ClientConfig{Servers: []string{"127.0.0.1"}, Port: fallbackPort}
+	srv, err := NewDNSServerWithConfig(config, zone, nil, port)
+	wt.AssertNoErr(t, err)
+	defer srv.Stop()
+	go srv.Start()
+	time.Sleep(100 * time.Millisecond) // Allow sever goroutine to start
+
+	// Create a regular UDP client and a TCP client
+	uc := new(dns.Client)
+	uc.UDPSize = UDPBufSize
+	tc := new(dns.Client)
+	tc.Net = "tcp"
+
+	m := new(dns.Msg)
+	m.SetQuestion(localTestName, dns.TypeA)
+	m.RecursionDesired = false
+
+	Debug.Printf("Using UDP for quering about %s", localTestName)
+	r, _, err := uc.Exchange(m, dnsAddr)
+	wt.AssertNoErr(t, err)
+	if len(r.Answer) > 1 { // TODO: remove this when we have proper support for multiple responses (zettio/weave#338)
+		wt.AssertStatus(t, r.Rcode, dns.RcodeBadTrunc, "DNS response code")
+		if len(r.Answer) == ansCount {
+			wt.Fatalf(t, "%d answers received, but we expected less than that", len(r.Answer))
+		}
+	} else {
+		Debug.Println("... multiple reponses not supported (yet)")
+		wt.AssertStatus(t, r.Rcode, dns.RcodeSuccess, "DNS response code")
+	}
+
+	Debug.Println("Using TCP for the same query")
+	tr, _, err := tc.Exchange(m, dnsAddr)
+	wt.AssertNoErr(t, err)
+	wt.AssertStatus(t, tr.Rcode, dns.RcodeSuccess, "DNS response code (TCP)")
+	if len(r.Answer) > 1 { // TODO: remove this when we have proper support for multiple responses (zettio/weave#338)
+		wt.AssertEqualInt(t, len(tr.Answer), ansCount, "Number of answers (TCP)")
+	}
+	wt.AssertType(t, tr.Answer[0], (*dns.A)(nil), "DNS record (TCP)") // assume they are all the same...
```

Comment:
Hmmm, not sure about that... the point is to check that, after a truncated response with UDP, you can receive the full response with TCP, so testing that you can receive whatever you send with TCP does not make much sense (that would only test that DNS with TCP is reliable, something we already know/assume)